src/Utilities.js

```diff
@@ -0,0 +1,123 @@
+import * as CESymbols from './CustomElementInternalSymbols';
+
+const reservedTagList = new Set([
+  'annotation-xml',
+  'color-profile',
+  'font-face',
+  'font-face-src',
+  'font-face-uri',
+  'font-face-format',
+  'font-face-name',
+  'missing-glyph',
+]);
+
+/**
+ * @param {string} localName
+ * @returns {boolean}
+ */
+export function isValidCustomElementName(localName) {
+  const reserved = reservedTagList.has(localName);
+  const validForm = /^[a-z][.0-9_a-z]*-[\-.0-9_a-z]*$/.test(localName);
+  return !reserved && validForm;
+}
+
+/**
+ * @private
+ * @param {!Node} node
+ */
+export function isConnected(node) {
+  // Use `Node#isConnected`, if defined.
+  const nativeValue = node.isConnected;
```

Comment:
(from offline) Native `isConnected` returns true if the element has a deep root that is *any* document, but `{dis}connectedCallback` should only be called if the element moves in and out of (the main document and html import documents of connected links (?)). Relevant documents need to be marked when crawling in `upgradeTree`.